WIP: New csk reader

attom opened the pull request:
This is a WIP and I anticipate needing several rounds of revision to get consistent with draco best-practices. My goal is to have this and downstream changes merged before the next code release.

### Background

* Currently, draco contains `src/compton/Compton.cc`, which is a wrapper around CSK functionality
* CSK as a library can read csk ASCII files and perform interpolation in temperature
* CSK also has exe to produce csk ASCII files

### Purpose of Pull Request

* CSK was built to be correct, not performant
* Reading ASCII files and interpolating are slow and memory-intensive in CSK
* This is the first step toward replacing Compton and the need to link to CSK
* Much like draco contains a reader/interpolator for opacity data, this adds functionality to read/interpolate Compton scattering data

### Description of changes

* Add preliminary reader/interpolator for Compton scattering data in `src/compton2`
* Compton2 class reads binary csk file and interpolates in temperature
* Compton2 is significantly faster than Compton (CSK)
* Added exe (CSKrw) to convert csk ASCII to binary, removing unneeded Legendre moments (for out_lin, out_nonlin, in_nonlin) and pre-computing temperature derivatives
* Added tests for CSKrw and Compton2
* This represents only a preliminary functionality; more will be added in the future
* For now, keep both Compton and Compton2

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [ ] Travis/Appveyor CI checks pass
  * [ ] Code coverage does not decrease
  * [ ] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation

KineticTheory commented:
@attom (1) I can help you with the Appveyor build issues. (2) Ignore the "Lint Code Base" failure.  (3) You should be able to fix the Travis failures for autodoc and style with relative ease.

I also want to help you clean up that long history so we don't see it in the merged 'develop' branch history.  I've been meaning to provide some guidance on this problem to other developers.  Let me work through the procedure write up and and I'll have you test it.

KineticTheory commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+    std::cout << "Testing interpolation at T = " << (interp_T * mec2) << " keV\n";
+
+    std::vector<std::vector<std::vector<std::vector<double>>>> interp_data =
+        compton_test->interpolate_csk(interp_T);
```
At some point, we should rethink this storage type.  This creates a lot of indirection.  (no change required for this PR.)

attom replied:
Oh yeah. This exact point is a major reason why we're moving from compton (which uses vector<vector<vector<vector<double>>>>) to compton2 (which uses vector<double>)

KineticTheory commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+        2.875946793361e-23, 1.265594484598e-25, 4.274264612141e-25, -9.142968009692e-27, 0, 0, 0, 0,
+        5.392751048597e-25, -1.344066881751e-26, 8.87128629256e-25, 3.986007245261e-27};
+
+    ut.check(std::equal(flat_interp_data.begin(), flat_interp_data.end(), flat_interp_gold.begin(),
```
simplify with `soft_equiv(a.begin(), a.end(), b.begin(), ...)`

attom replied:
what's the last argument? the tolerance?

KineticTheory replied:
> what's the last argument? the tolerance?

Yes. 

```c++
template <typename Value_Iterator, typename Ref_Iterator>
inline bool
soft_equiv(Value_Iterator value, Value_Iterator value_end, Ref_Iterator ref, Ref_Iterator ref_end,
           typename std::iterator_traits<Value_Iterator>::value_type const precision = 1.0e-12);
```

or 

```c++
inline bool soft_equiv(const std::vector<double> &value, const std::vector<double> &ref,
                       const double precision = 1.0e-12
```

KineticTheory commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+    ut.check(sz3 == num_leg_moments_gold, "tested Legendre moments size");
+
+    // Print result (useful if golds need updating)
+    if (do_print) {
```
Since you create/use this in more than one place, consider making this a helper function that is only defined at the top of this source file.

attom replied:
will do

KineticTheory commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+    std::cout << "CSK exception thrown. " << std::endl;
+    // We successfully caught the bad file!
+    caught = true;
+  }
```
You might also want the `catch(...)` case. You can omit it here and just let the top level catch work its magic.

attom replied:
Is it just replace the catch statement with " catch(...)" or does something special need to replace the ellipses. I'm not overly familiar with this language feature or exceptions in general

KineticTheory replied:
> Is it just replace the catch statement with " catch(...)" or does something special need to replace the ellipses. I'm not overly familiar with this language feature or exceptions in general

Since this is is a test, I think `catch(...)` is appropriate.  You can fine tune the error message or recovery behavior with a more detailed `catch(arg)`, if there is need to do so.

KineticTheory commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+}
+
+//------------------------------------------------------------------------------------------------//
+// End of test/tCompton.cc
```
`t2Compton.cc`

attom replied:
good catch

KineticTheory commented on src/compton2/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+#-----------------------------*-cmake-*----------------------------------------#
+# file   src/compton/CMakeLists.txt
+# author Andrew Till
```
I think merging this into `compton` is best.  Let me know what you think.

attom replied:
I'd like to keep them separate so that we could remove the entire `compton` directory later if we wanted. Otherwise, we'd have to be very careful about what was a standalone functionality and what was a wrapper.

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+//----------------------------------*-C++-*-----------------------------------//
+/*!
+ * \file   compton2/Compton2.cc
```
As I mentioned earlier, let's find something more descriptive than `2`.

attom replied:
yep

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+ * \date   11 May 2020
+ * \brief  Implementation file for compton interface
+ * \note   Copyright (C) 2017-2020 Triad National Security, LLC.
+ *         All rights reserved. */
```
Same comments as before wrt 100-col format.

attom replied:
will update

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iomanip> // RM
```
Did you mean to remove some of these include directives?

attom replied:
yeah, they were used while I was debugging. good catch

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  UINT index = (loc - xs.cbegin());
+
+  // Move index and x to be interior
+  // len >=2 assumed
```
Add DbC to for `len >=2`.

attom replied:
good call

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+
+namespace rtt_compton2 {
+
+// xs is assumed to be monotonically increasing and unique
```
Please add doxygen style comment block.
Did you intend for this to be a file-scope free function?
Would it make more sense to define these as template functions `<UINT,FP>` and omit the `#define` above?

attom replied:
I'll dig around for one. Yes, the function is a helper function that doesn't need to be in the class, though I'm not sure who else would use it.

We could templatize on UINT and FP. Were you thinking just for the free functions or for the entire class?

KineticTheory replied:
> I'll dig around for one. Yes, the function is a helper function that doesn't need to be in the class, though I'm not sure who else would use it.
> 
> We could templatize on UINT and FP. Were you thinking just for the free functions or for the entire class?

I was thinking for the entire class since you use these CPP macros in a way that a native C++ template could be used.  The template form is less error prone and the static analysis tools do a better job of checking.

attom replied:
I need to think more about this. I'd like FP and UINT to be the same as the data types that are stored in the binary csk data file but then I'd like to return doubles and size_t or uint64_t's. Does that seem more reasonable?

KineticTheory replied:
That makes perfect sense.  CSK should be able to publish the data types through a header or `config.h` or even through the cmake system. (`csk-config.cmake`).  You might even be able to use to C++'s [decltype](https://en.cppreference.com/w/cpp/language/decltype) to return the type of a data field defined by CSK.

```c++
class compton2 {
private:
  // DATA
  decltype(csk::fp_type) csk_data;
  double compton2_data;
...}
```

I'm not sure if that snippet is valid C++, but you get the idea.

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  return H;
+}
+
+Compton2::Compton2(std::string filename)
```
Add Class doc block.

attom replied:
I'll dig around for one. Is anyone using Doxygen these days?

KineticTheory replied:
* Try using one from `ds++/DracoStrings.hh` 
* Something like:
```c++
//------------------------------------------------------------------------------------------------//
/*!
 * \brief one-liner
 *
 * \param[in] string_in This string will be converted letter by letter to lowercase.
 * \return A string that contains no uppercase letters.
 *
 * Longer description as needed, with design, intent, possibly examples.
 */
```

KineticTheory commented on src/CMakeLists.txt:
```diff
@@ -97,6 +97,7 @@ message(" ")
 message( STATUS "Configuring Level 5 packages --" )
 add_dir_if_exists( mesh )       # needs c4, ds++, mesh_element
 add_dir_if_exists( quadrature ) # needs mesh_element, parser, special_functions
```
This PR needs to be checked with clang-tidy.  There are many simplifications that should be implemented, but it easier to iterate with the tool than for me to list them in this review.  I'll send instructions later.

attom replied:
Thanks! Should I do all these edits on the squashed branch or the current branch?

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  constexpr int bcast_rank = 0;
+
+  // Broadcast sizes
+  UINT data_size = data_.size();
```
This is confusing to me.  Maybe use `auto`?  This type will always be `size_t`, but `size_t` isn't always the same as `unsigned`.

attom replied:
I was hoping to get your help more generally on the broadcast / receive and read functions to be sure they're architecture (Endian-ness, etc.) portable.

Perhaps we could have a WebEx on what to do for these?

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  data_size = pack[p++];
+
+  // Derived sizes
+  const auto tsz = int(num_temperatures_);
```
Avoid c-style casts.

attom replied:
Do you prefer static_cast<int>?

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  // Broadcast grids
+  if (rank != bcast_rank)
+    Ts_.resize(tsz);
+  rtt_c4::broadcast(&Ts_[0], tsz, 0);
```
Avoid using `&vec[0]` without extra precautions as this is undefined behavior if the vector is empty.

attom replied:
See above comment. Not sure the best way to do the broadcast for a vector.

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  rtt_c4::broadcast(&derivs_[0], dsz, 0);
+}
+
+int Compton2::read_binary(std::string filename) {
```
Pass strings by `const &` unless you intend to use `std::move`.

attom replied:
Good best-practices. Will do.

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+
+int Compton2::read_binary(std::string filename) {
+  // Read
+  auto fin = std::ifstream(filename, std::ios::in | std::ios::binary);
```
Check for failure?

attom replied:
Do I use `fin.good()` for that? And throw if not good?

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  auto fin = std::ifstream(filename, std::ios::in | std::ios::binary);
+
+  // Ensure valid type
+  char expected[] = " csk ";
```
Clang-tidy will require you to change some of this.

attom replied:
The binary file is a set of chars that starts with " csk ". How should I compare to the start of the file if not char[]?

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+    return 1;
+  }
+
+  UINT binary_ordering, version_major, version_minor;
```
One line per declaration with default initialization.

attom replied:
sure

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  fin.read(reinterpret_cast<char *>(&version_major), sizeof(UINT));
+  fin.read(reinterpret_cast<char *>(&version_minor), sizeof(UINT));
+  fin.read(reinterpret_cast<char *>(&binary_ordering), sizeof(UINT));
+  if (version_major != 1 || binary_ordering > 1) {
```
Check for read errors or state of the istream?
It would likely be more efficient to read the whole file into a buffer and then parse the buffer?

attom replied:
I defer to you. Should one rank read the file, put it in a buffer, and then MPI communicate that buffer to the other ranks and have everyone parse?

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  num_points_ = num_evals_ + num_leg_moments_ - 1U;
+  UINT egsz = gsz + 1;
+
+  if (false) {
```
Remove or add a comment.  Consider `#if 0` to keep this out of the binary.

attom replied:
Yeah, this is a debug print statement. Do you think there's value in making a function that prints the contents of a file or should I just remove this (or keep around with `#if 0`)?

KineticTheory replied:
I really think that decision is up to you.  In more mature code, I would recommend that the code be removed entirely. Since this is brand new code, it might be worth keeping it in place locally.  Optionally, you could create a utility application that reads the file and dumps the contents to the screen.  We have something similar for both cdi_ipcress and cdi_eospac.

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  return 0;
+}
+
+////////////////////////////////////////////////////////////////////////////////
```
Please use standard draco formatting.

attom replied:
Must have been a copy-paste from somewhere. Is standard draco formatting `/* ---- ... ---- */`?

KineticTheory commented on src/compton2/Compton2.cc:
```diff
@@ -0,0 +1,475 @@
+  UINT iT = rtt_compton2::find_index(Ts_, Teff);
+
+  // Fill Hermite function
+  std::array<FP, 4> H = rtt_compton2::hermite(Teff, Ts_[iT], Ts_[iT + 1U]);
```
Avoid the use of single letter variables except for look indices.  It is difficult to search for such variables.

attom replied:
Will change to `hermite_eval` or similar. Preferences on camel case vs underscore?

KineticTheory replied:
Historically, our developers prefer the underscore separator.

KineticTheory commented on src/compton2/Compton2.hh:
```diff
@@ -0,0 +1,197 @@
+//----------------------------------*-C++-*-----------------------------------//
+/*!
```
Use standard 100-col format (I know - we adopted this format after you started this branch. sorry.)

attom replied:
should be easy enough to find a place to copy-paste from

KineticTheory commented on src/compton2/Compton2.hh:
```diff
@@ -0,0 +1,197 @@
+#include <string>
+#include <vector>
+// Draco libraries
+#include "compton2/Sparse_Compton_Matrix.hh"
```
Remove this comment and let clang-format order these include directives.  The stdlib headers should be listed after draco headers.

attom replied:
I'll check the squashed version to see what git-format did

KineticTheory commented on src/compton2/Compton2.hh:
```diff
@@ -0,0 +1,197 @@
+
+namespace rtt_compton2 {
+
+enum Eval : size_t { in_lin = 0, out_nonlin = 1, nl_diff = 2 };
```
prefer enum class.

attom replied:
will change

KineticTheory reviewed (changes requested):
This isn't a full review, but I think it will get you started on some of the changes I'm requesting.  I'll update my notes after a 2nd pass.

keadyk commented:
Woot!! Nice, @attom! Let me know if you want more eyes on this (I don't want to jump the gun if you're still working on it :) )

attom commented on src/CMakeLists.txt:
```diff
@@ -97,6 +97,7 @@ message(" ")
 message( STATUS "Configuring Level 5 packages --" )
 add_dir_if_exists( mesh )       # needs c4, ds++, mesh_element
 add_dir_if_exists( quadrature ) # needs mesh_element, parser, special_functions
+add_dir_if_exists( compton2 )     # needs ds++, units, cdi_analtyic
```
I like the idea of renaming both, though my initial goal was to not break anything that depended on `compton`. If I rename the directory, will CMakeLists.txt in CTA/TRT/Jayenne be smart enough to know where to look?

attom replied:
Thoughts on what to name the directories? I like the idea of having them separate, because they do very different things. `compton` wraps an interface to `csk` while `compton2` handles reading / interpolating csk data files itself.

`compton` -> `compton_wrapper`
`compton2` -> `compton_tools`

?

warsa replied:
How about csk_interface or compton_interface for the "wrapper" part, compton_data for the data "reading/handling" part?

KineticTheory replied:
> I like the idea of renaming both, though my initial goal was to not break anything that depended on `compton`. If I rename the directory, will CMakeLists.txt in CTA/TRT/Jayenne be smart enough to know where to look?

Draco's build system publishes all of the path and library information automatically.  In cta, you will need to tell the code to link the libraries `compton_interface` and `csk_interface` (or whatever names we end up with).  Any C++ source files that currently has `#include compton/Compton.hh` will need to be updated.  This can be done easily with a sed command or through an IDE _refactor_ command.

attom replied:
Alright, time for a vote: 

`compton` -> `compton_interface`

or

`compton` -> `compton_wrapper`

@warsa or @keadyk or @KineticTheory ?

I'm planning on using `compton2` -> `compton_tools` unless there is an objection

attom replied:
I'm planning on using `compton_interface` unless there is a major objection

keadyk replied:
Sounds good to me! :)

warsa replied:
I prefer it as well.

attom commented on src/compton/test/CMakeLists.txt:
```diff
@@ -45,6 +48,13 @@ else()
+  add_parallel_tests(
+    SOURCES "${test2_sources}"
+    PE_LIST "1"
+    DEPS    "Lib_compton;Lib_c4"
```
Will try without

attom commented on src/compton/test/CMakeLists.txt:
```diff
@@ -16,6 +16,9 @@ project( compton_test CXX )
   ${PROJECT_SOURCE_DIR}/tCompton.cc )
 
+set( test2_sources
+  ${PROJECT_SOURCE_DIR}/t2Compton.cc )
```
Agreed. Will update. This test uses dummy data, so maybe include that in the name?

KineticTheory replied:
> Agreed. Will update. This test uses dummy data, so maybe include that in the name?

That is the pattern I've used for some of the `cdi_*` tests, but I'm not too picky about names in test directories.

attom commented on src/compton/test/CMakeLists.txt:
```diff
@@ -45,6 +48,13 @@ else()
+    SOURCES "${test2_sources}"
+    PE_LIST "1"
+    DEPS    "Lib_compton;Lib_c4"
+    LABEL   "nomemcheck"
```
Not that I'm aware of. Probably a copy-paste error from wherever I got the CMake. Will try removing

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+//-----------------------------------*-C++-*----------------------------------//
```
I'll look around for one

KineticTheory replied:
Look in `ds++`.  This is a new change in Draco and we are still updating some of the files that didn't get fixed by my one-pass sed script.  If you use Draco's Emacs helpers, the updated dividers are available from within Emacs via key bindings.   `M-x draco-mode`, followed by `M-x describe-mode` and scrolling down to the section on 'draco minor mode'.

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+ * \author Andrew Till
+ * \date   2020 Oct 14
+ * \brief  Implementation file for t2Compton
+ * \note   Copyright (C) 2017-2020 Triad National Security, LLC.
```
Will update

attom replied:
Where do we want the ending `*/`? Same line or next line down or does it not matter?

KineticTheory replied:
Prefer the same line as the copyright.

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+            << "---------------------------------------------------------\n";
+
+  // open a small mg opacity file:
+  const std::string filename = ut.getTestSourcePath() + "../../compton2/test/dummy_data";
```
I wanted to be sure that compton and compton2 were getting the same answer for the same input, which is why I wrote this test. They are. Since compton is already tested, should I just remove this test? It's using the data in compton2 but the methods in compton.

KineticTheory replied:
You can leave it here.  I just wanted to make sure that this design was intentional.

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+    FAILMSG("Failed to construct a Compton object!");
+    // if construction fails, there is no reason to continue testing...
+    return;
+  }
```
Can do. This is a copy-paste from `src/compton/test/tCompton.cc`

KineticTheory replied:
> Can do. This is a copy-paste from `src/compton/test/tCompton.cc`

Again, this can be left here if you are okay with the test aborting immediately if some other exception is thrown.

attom replied:
Yes, that's fine with me

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+
+  // Unitless (divided by mec2)
+  // NB: These values can be read directly from the ASCII data files
+  // (3rd line)
```
sorry, 3rd line from the csk data file. I'll edit the comment to be more explicit.

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+  line2_Ts_gold[1] *= mec2;
+  */
+
+  Ensure(grp_bds.size() == (num_groups_gold + 1U));
```
Yeah, another copy-paste from `src/compton/test`. I'll fix.

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+  */
+
+  Ensure(grp_bds.size() == (num_groups_gold + 1U));
+  ut.check(std::equal(grp_bds.begin(), grp_bds.end(), grp_bds_gold.begin(),
```
Oh, nifty! Is that `soft_equiv_deep`?

KineticTheory replied:
> Oh, nifty! Is that `soft_equiv_deep`?

That is the actual implementation. There are specializations for 1, 2 and 3D `std::vector` (or you can add your own).  See https://github.com/lanl/Draco/blob/dd14663b05f77866335596b3c52d39fdd8964a1b/src/ds%2B%2B/Soft_Equivalence.hh#L204 and https://github.com/lanl/Draco/blob/dd14663b05f77866335596b3c52d39fdd8964a1b/src/ds%2B%2B/test/tstSoft_Equiv.cc#L195

attom replied:
Excellent. I'll have to add this to my other unit tests in consuming codes

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+                      [tol](double a, double b) -> bool { return soft_equiv(a, b, tol); }),
+           "checked group boundaries");
+
+  Ensure(T_evals.size() == num_T_evals_gold);
```
Will fix.

attom commented on src/compton/test/t2Compton.cc:
```diff
@@ -0,0 +1,322 @@
+           "checked temperature grid");
+
+  if (!soft_equiv(compton_test->get_min_etemp(), line2_Ts_gold[0]))
+    FAILMSG("Min etemp read incorrectly!");
```
I wish I knew the Draco testing macros / functions better. Maybe we should have a review meeting for Draco developers?

KineticTheory replied:
> I wish I knew the Draco testing macros / functions better. Maybe we should have a review meeting for Draco developers?

That is probably a good thing to do.  I just takes a lot of energy that I seem to be in short of lately. 😄

attom commented:
Howdy @keadyk. Yes, another pair of eyes would be most appreciated, but perhaps after my next commit, since there's a lot to be done. While I have you, what do you think about naming? @KineticTheory and I are debating how to name the directories, currently the unhelpful `compton` and `compton2`, or whether they should be merged into one directory, which I think would be more confusing than helpful. Thoughts?

KineticTheory commented:
@attom I hope I didn't discourage you with that long list of comments.  It is great to see these new capabilities going into the code base!  Let me know how else I can help.

keadyk commented:
@attom I agree with you -- I think it would be hard to parse if everything were in the same directory. Since we will likely scrap the original compton folder in the future, were you thinking of just renaming compton2 -> compton at that point? :)

Hmmm. Maybe we could go with something like csk_compton and fast_compton? It seems like adding "_reader" to the end of the directory name would be too unwieldy, and not entirely accurate. *ponders*

attom commented:
What do you think of `compton_wrapper` (for `compton`) and `compton_tools` (for `compton2`)? Or `compton` -> `csk_wrapper`? Or `compton` -> `csk_compton` as you suggest. The difficulty is I've been using csk for both the library and the file format.

keadyk commented:
Ohh, okay, I see! Let me think about this a little more. My first impression is that I think I like the "compton_tools" name for your new code. :)

keadyk commented:
@attom Ok, I think the compton_wrapper / compton_tools naming scheme makes the most sense to me. I also use "csk" to refer to the library, the data, the actual expression for the kernel, etc., so that's probably not a helpful way to differentiate them (as you pointed out) :)

attom commented:
Alright, I've addressed most of the comments above. Big things that I haven't addressed that will need more thought:

* FP and UINT: should we template on these? how do we divorce the underlying data format of the binary file from the interface to the class?
* broadcasting: best practices for broadcasting a vector?
* binary read: best practices for reading a vector of unsigned 64-bit ints or doubles from a character stream in a way that is correct for multiple architectures?
* comparing an array of chars from a binary file to a known string; should I use std::array<char, length> and std::equal?
* how to use clang-tidy on this
* I'll resubmit as a PR with a single commit once we're closer to merging

Thanks @KineticTheory for guiding me through this so far! If anyone else has concerns, I think we're getting close to a good time for y'all to voice them

KineticTheory commented:
> * binary read: best practices for reading a vector of unsigned 64-bit ints or doubles from a character stream in a way that is correct for multiple architectures?

This is straight forward and there are a ton of examples on line.  Let me know if you need any help.  I've done this a few times, but I don't think any of that code is in Draco.  The main thing to avoid for portability is **Fortran**.  Fortran does non-portable padding to the header/footer blocks with binary files.  I can't remember how _Endianess_ plays with binary data.  However, `ds++/Endian` provides some help if you need to swap data around.

> * comparing an array of chars from a binary file to a known string; should I use std::array<char, length> and std::equal?

* I would probably ready the data in binary and stuff it into a `std::string` and use simple `==`  checks.  When saving strings to a binary file, you normally save the field size and then the data.  That way you know how many bytes to load and stuff into the string container.  `ds++/DracoStrings` may have some useful functions to cast raw data into strings and other types.

> * how to use clang-tidy on this

```console
ccscs% module swap gcc llvm
ccscs% cmake C_FLAGS=-Werror CXX_FLAGS=-Werror cmake -DDRACO_STATIC_ANALYZER=clang-tidy ~/draco
ccscs% make -j 20 -k  # build as much as you can before running with VERBOSE=1
ccscs% make VERBOSE=1
```

* Then use http://clang.llvm.org/extra/clang-tidy/checks/list.html to understand why clang-tidy is complaining.

KineticTheory commented:
When you are looking at *clang-tidy*, let me know if there are additional rules that you think should be enabled by default.

attom commented:
Whew. Applied clang-tidy to both `compton_interface` and `compton_tools`. Surprisingly, fewer changes than I thought, which might just mean we need to add more checks to clang-tidy.

Tomorrow, I hope to look at portability (Endianness, etc.)

attom commented:
@KineticTheory, do we have a system with non-standard endianness for testing? Or a system that isn't 64 bit in the normal way?

KineticTheory commented:
> @KineticTheory, do we have a system with non-standard endianness for testing? Or a system that isn't 64 bit in the normal way?

Actually, I don't think we have any big-endian machines right now.  Sequoia was big-endian but that machine is gone.  Let's punt on this concern for now.

KineticTheory commented on src/compton_tools/python/common_compton.py:
```diff
@@ -1,22 +1,23 @@
 #!/usr/bin/env python
-#---------------------------------------*-python-*-------------------------------------------------#
+# --------------------------------------*-python-*------------------------------------------------ #
```
Did the linter really want those extra spaces?  I guess I need to update the Emacs template to match.

attom commented:
Open questions:

* Read the whole file into a buffer and then parse the buffer vs current strategy of read-as-you-go (can we defer this for a future merge? I'm not quite sure the mechanics of the former option)

* Is broadcasting okay as it stands for now? I've added checks for sizes.

* FP and UINT; not sure what I want or what the best thing to do is. Goals are:
  * Have interface return doubles, since that's what the rest of our codes use for floating-point types
  * Read and write to the binary file using FP (currently double) and UINT (currently uint64_t) for consistency
  * Safely convert as needed from FP to doubles using static_cast<double>(FP val); I assume this is currently a no-op?
  * Question: Should the conversion happen during the read phase or the interpolate phase? (@KineticTheory or @keadyk, thoughts?)

keadyk commented:
@attom I would think it'd be easiest to convert when you read (if I understand correctly!), but I don't have a strong opinion :)

attom commented:
That works for me. I'll play around with the change tomorrow.

attom commented:
Alright, finally getting back around to this. @keadyk and @KineticTheory could you please take another look? Once we're happy, I'll close this PR and make a new branch with just one commit of the changes

keadyk commented:
I admittedly did some skimming, but what I've seen looks reasonable to me! Excited to see it in action :)

KineticTheory commented:
> Alright, finally getting back around to this. @keadyk and @KineticTheory could you please take another look? Once we're happy, I'll close this PR and make a new branch with just one commit of the changes

I should have some time this evening or tomorrow afternoon to give this another review.  Sorry for the delay.

KineticTheory commented on src/compton_tools/Sparse_Compton_Matrix.hh:
```diff
@@ -0,0 +1,101 @@
+ *        matrix multiplication; can store an arbitrary number of points,
+ *        where each point represents an  M x G x G matrix (moments, groups)
+ *
+ *        THIS CLASS IS A STUB (not yet fully implemented or used)
```
If this isn't needed, then maybe is shouldn't be in the PR?  Or maybe you can leave it in a `#ifdef 0` block for now.

attom replied:
Yeah, I'll add `#ifdef 0`

KineticTheory commented on src/compton_tools/cskrw.cc:
```diff
@@ -0,0 +1,1023 @@
+  std::vector<FP> derivs;
+};
+
+Sparse_Compton_Data::Sparse_Compton_Data() : first_groups(), indexes(), data(), derivs() {}
```
I think this can be removed.

attom replied:
replaced with default constructor

KineticTheory commented on src/compton_tools/cskrw.cc:
```diff
@@ -0,0 +1,1023 @@
+
+  // VALUES USED IN CSK'S PHYSICAL_CONSTANTS.HH
+  // Normalization for energy and nonlinear terms
+  const FP mec2 = 510.998;
```
* Probably better to pull these values directly from csk.
* This file looks a bit rough, maybe just add some history and discussion to the top of the file.  It doesn't appear to be something that would be consumed by users.

attom replied:
I'd like to have a code that in no way requires the CSK library. Added history and intent comments.

KineticTheory commented on src/compton_tools/python/common_compton.py:
```diff
@@ -0,0 +1,89 @@
+
+
+# ------------------------------------------------------------------------------------------------ #
+def print_grids(grids, fileroot, verbosity=False):
```
@alexrlongne Can you review the python here?

attom replied:
For reference, all of the python in `src/compton_tools/python` are little scripts intended for internal use so we don't have to keep rewriting our own python readers of Compton data

alexrlongne replied:
Looks good, just the small comment about the normalization constant.

KineticTheory reviewed (changes requested):
This is looking much more mature.  Thanks for working on it. 

I still want to look at the code coverage numbers before approving.

attom commented on .clang-format:
```diff
@@ -61,7 +61,7 @@ ForEachMacros:
 IncludeBlocks:   Preserve
 IncludeCategories:
-  - Regex:           '^"(c4|cdi|cdi_analytic|cdi_CPEloss|cdi_eospac|cdi_ipcress|cdi_ndi|compton|device|diagnostics|ds\+\+|experimental|fit|linear|memory|mesh|mesh_element|meshReaders|min|norms|ode|parser|plot2D|quadrature|rng|roots|RTT_Format_Reader|special_functions|timestep|units|viz)/'
+  - Regex:           '^"(c4|cdi|cdi_analytic|cdi_CPEloss|cdi_eospac|cdi_ipcress|cdi_ndi|compton_interface|compton_tools|device|diagnostics|ds\+\+|experimental|fit|linear|memory|mesh|mesh_element|meshReaders|min|norms|ode|parser|plot2D|quadrature|rng|roots|RTT_Format_Reader|special_functions|timestep|units|viz)/'
```
Done (will appear in next commit)

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+ * \file   compton_tools/Compton_Native.cc
+ * \author Andrew Till
+ * \date   11 May 2020
+ * \brief  Implementation file for native compton bindary-read and temperature interpolation
```
Fixed

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+ * \brief find location in a sorted list and min/max value to be within list
+ *
+ * \param[in] xs vector of monotonically increasing and unique values
+ * \param[inout] x value whose location in xs needs to be found;
```
Done

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  size_t index = (loc - xs.cbegin());
+
+  // Move index and x to be interior
+  size_t len = xs.size();
```
Updated

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  size_t len = xs.size();
+  double xmin = xs[0];
+  double xmax = xs[len - 1U];
+  index = std::min(len - 1U, std::max(static_cast<size_t>(1), index)) - 1U;
```
Yes, it was unavoidable. Otherwise, get `no matching function for call to 'max(size_t&, unsigned int)'`

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+ * interpolates the value of the function at x, where this function computes and returns H
+ */
+std::array<double, 4> hermite(double x, double xL, double xR) {
+  Require(xL <= x && x <= xR && xL < xR);
```
Makes sense. Fixed.

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+ * H[0] * vL + H[1] * vR + H[2] * dL + H[3] * dR
+ * interpolates the value of the function at x, where this function computes and returns H
+ */
+std::array<double, 4> hermite(double x, double xL, double xR) {
```
I'm not going to move it until we know where it's going

attom replied:
I templated it on floating-point type

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  double dx = xR - xL;
+  // Left/right basis functions
+  double bL = (xR - x) / dx;
+  double bR = (x - xL) / dx;
```
Fixed

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+Compton_Native::Compton_Native(const std::string &filename)
+    : num_temperatures_(0U), num_groups_(0U), num_leg_moments_(0U), num_evals_(0U), num_points_(0U),
+      Ts_(0U), Egs_(0U), first_groups_(0U), indexes_(0U), data_(0U), derivs_(0U) {
+  Require(filename.length() > 0U);
```
yes, we tend to use trailing underscores for private member data functions

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  size_t data_size = data_.size();
+  std::array<size_t, 6> pack = {num_temperatures_, num_groups_, num_leg_moments_,
+                                num_evals_,        num_points_, data_size};
+  rtt_c4::broadcast(&pack[0], pack.size(), bcast_rank);
```
pack is an `array` and guaranteed to be size 6, right?

keadyk replied:
I agree with @attom, it's a std array, so should be sized at compile time to 6 - always. I don't think the reference to element 0 should be an issue in this context :)

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  const auto egsz = static_cast<int>(num_groups_ + 1U);
+  const auto fgsz = static_cast<int>(num_temperatures_ * num_groups_);
+  const auto isz = static_cast<int>(fgsz + 1U);
+  const auto dsz = static_cast<int>(data_size);
```
I think I needed int's for Draco's MPI broadcast to work? Should never exceed 2 billion data points in practice (if it did, would take 30+ GB for data and derivatives). I'll add a check on data_size to insist it is less than `1U << 31U`

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  if (rank != bcast_rank)
+    Ts_.resize(tsz);
+  Check(Ts_.size() == static_cast<size_t>(tsz));
+  rtt_c4::broadcast(&Ts_[0], tsz, 0);
```
Will the broadcast correctly fail with nullptr?

keadyk replied:
It *should* just be a pass-through in that case, because it'll get passed a size of zero via tsz :)

keadyk replied:
(But I believe @KineticTheory 's original comment about the "&" syntax still stands)

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  }
+
+  // Ensure valid type
+  // (using vector of char to avoid using C-style strings)
```
Good idea. Done

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+    for (char c : actual)
+      std::cerr << c;
+    std::cerr << "'" << std::endl;
+    return 2;
```
Yes, it's so that all of the ranks can die together if rank 0 has a problem with the read

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  for (size_t i = 0; i < tsz; ++i) {
+    // Convert from FP (type in the binary file) to double
+    FP tmp;
+    fin.read(reinterpret_cast<char *>(&tmp), sizeof(FP));
```
Right now, both Compton_Native.cc (data reader) and cskrw.cc (data writer) define FP and UINT as the same types, which is required by the file format that is in the software requirements document. Do you want me to make a header file that defines these types and then have both cc files include the header?

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+
+#if 0
+    std::cout << '\n';
+    std::cout << "-------------------------------------------------\n";
```
Yes, perhaps when the code matures and stabilizes more

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+/*!
+ * \brief Interpolate csk data in temperature and return dense linear inscattering matrix
+ *
+ * \param[inout] inscat The flattened (1D), dense inscattering matrix as a vector
```
Fixed

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+ */
+void Compton_Native::interp_dense_inscat(vec_d &inscat, double Te_keV,
+                                         size_t num_moments_truncate) const {
+  // Finds index and nudges Teff such that
```
For robustness, I don't require Te_keV > 0. The `find_index` function deals with extrapolation concerns gracefully. What specifically would you like to see checked by DbC?

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  const size_t eval_offset = 0; // in_lin
+
+  // Resize and fill with zeros
+  inscat.resize(end_leg * num_groups_ * num_groups_);
```
Wouldn't be the same. `resize` won't overwrite preexisting values and I want to be sure inscat is filled with zeros everywhere (inscat is passed in as a function argument and so could have who knows what)

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+  std::fill(inscat.begin(), inscat.end(), 0.0);
+
+  // Apply Hermite function
+  for (size_t k = 0; k < end_leg; ++k) {
```
Yes, I studied the loop ordering in my stand-alone test code and found this was the fastest on x86 procs

attom commented on src/compton_tools/Compton_Native.hh:
```diff
@@ -0,0 +1,220 @@
+ * \author Andrew Till
+ * \date   11 May 2020
+ * \brief  Implementation file for native compton bindary-read and temperature interpolation
+ * \note   Copyright (C) 2017-2020 Triad National Security, LLC. All rights reserved.
```
Fixed in several places

attom commented on src/compton_tools/Compton_Native.hh:
```diff
@@ -0,0 +1,220 @@
+ * 1) access (interpolate) data from existing multigroup csk libraries
+ * 2) obtain auxiliary information from existing multigroup libraries
+ *    (electron temperature bounds, frequency group structures, etc)
+ *
```
fixed

attom commented on src/compton_tools/Compton_Native.hh:
```diff
@@ -0,0 +1,220 @@
+  //--------------------------------------------------------------------------//
+
+  // \brief Constructor
+  Compton_Native(const std::string &filename);
```
Added `explicit`. Probably disable?

attom commented on src/compton_tools/Sparse_Compton_Matrix.cc:
```diff
@@ -0,0 +1,50 @@
+#include "compton_tools/Sparse_Compton_Matrix.hh"
+#include "c4/global.hh"
+#include "ds++/Assert.hh"
+
```
Done

attom commented on src/compton_tools/Sparse_Compton_Matrix.cc:
```diff
@@ -0,0 +1,50 @@
+
+namespace rtt_compton_tools {
+
+Sparse_Compton_Matrix::Sparse_Compton_Matrix(size_t num_points, size_t num_groups,
```
This is a stub class at present that's commented out of CMakeLists.txt but which may see future action. I'll add a stub autodoc

attom commented on src/compton_tools/Compton_Native.hh:
```diff
@@ -0,0 +1,220 @@
+  // PRIVATE DATA
+  //--------------------------------------------------------------------------//
+
+  size_t num_temperatures_;
```
Added comments and changed to use default initialization

attom commented on src/compton_tools/Compton_Native.cc:
```diff
@@ -0,0 +1,632 @@
+
+  // Precompute some sparse indexes
+  const size_t sz = indexes_[indexes_.size() - 1];
+  const size_t end_leg = std::min(static_cast<size_t>(num_moments_truncate), num_leg_moments_);
```
It is. Removing

attom commented:
Alright. Ready for another round of reviewing

keadyk commented:
> Alright. Ready for another round of reviewing

*rings the bell* FIGHT! :)

I will take one more quick look, but I reckon this is about ready to be smooshed up and merged!

keadyk commented:
Alright, I went ahead and marked things that seemed cut-and-dried as "resolved" :)

@KineticTheory @attom On the interpolation functions front - I will gladly self-assign the task of moving Andrew's CSK-related (and, eventually, @brryan 's NDI-related) interpolation functions into a shared location in draco (@KineticTheory would adding an "interpolation" directory be okay?).

alexrlongne commented on src/compton_tools/python/csk_reader.py:
```diff
@@ -0,0 +1,138 @@
+
+    # normalizations
+    mec2 = 510.998  # keV
+    csk_norm = 2.0 * 0.037558  # currently approximate
```
@keadyk Mentioned that this is explained in CSK, do you want to include any of those comments here? Not a big deal.

attom replied:
Good call. Added comments and more digits

KineticTheory commented:
Replaced by #968
